File: JavaScript/binary-tree.js
// Suppose you're given a binary tree represented as an array.
//  For example, [3,6,2,9,-1,10] represents the following binary tree 
//  (where -1 is a non-existent node):

// Write a function that determines whether the left or right branch
// of the tree is larger. The size of each branch is the sum of the node values.
// The function should return the string "Right" if the right side is larger and "Left"
// if the left side is larger. If the tree has 0 nodes or if the size of the branches
// are equal, return the empty string.

// [3,6,2,9,-1,10] 
// Left

class TreeNode {
  constructor(value) {
      this.value = value;
      this.left = null;
      this.right = null;
  }
}

class Tree {
  constructor(head = null) {
      this.head = head;
  }

  setTree = (arr, i) => {
      let head = null;
      if (i < arr.length) {
          if (arr[i] === -1) { return head; }
          head = new TreeNode(arr[i]);
          head.left = new Tree().setTree(arr, i * 2 + 1);
          head.right = new Tree().setTree(arr, i * 2 + 2);
      }
      return head;
  }
  
  sumTree = (head) => {
      if (head == null) { return 0; }
      return head.value + new Tree().sumTree(head.left) + new Tree().sumTree(head.right);
  }
}

const solution = (arr) => {
  if (!arr || arr.length <= 1) { return ''; }
  const head = new Tree().setTree(arr, 0);
  const left = new Tree().sumTree(head.left);
  const right = new Tree().sumTree(head.right);
  console.log(left);
  console.log(right);
  return left === right ? '' : (left > right ? 'Left' : 'Right');
};

var a = [3,6,2,9,-1,10];
var result = solution(a);
console.log(result);

File: Swift/binary-tree.swift


class TreeNode {
  var val: Int
  var left: TreeNode?
  var right: TreeNode?

  init(_ val: Int) {
    self.val = val
    self.left = nil
    self.right = nil
  }
}

class Tree {
  var head: TreeNode?

  init(_ head: TreeNode? = nil) {
    self.head = head
  }

  func setTree(_ nums: inout [Int?], _ i: inout Int) -> TreeNode? {
    var root: TreeNode? = nil
    if i < nums.count {
      if let num = nums[i] {
        root = TreeNode(num)
        var left = i * 2 + 1
        var right = i * 2 + 2
        root?.left = setTree(&nums, &left)
        root?.right = setTree(&nums, &right)
      }
    }
    return root
  }

  fileprivate func setDict(_ root: inout TreeNode?, _ dict: inout [Int: Int], _ i: inout Int) {
    if let root = root {
      dict[i] = root.val
      var left = i * 2 + 1
      var right = i * 2 + 2
      setDict(&root.left, &dict, &left)
      setDict(&root.right, &dict, &right)
    }
  }

  func setArray(_ head: inout TreeNode?, _ i: inout Int) -> [Int] {
    var dict: [Int: Int] = [:]
    setDict(&head, &dict, &i)

    var arr: [Int] = []
    for i in 0..<dict.keys.count {
      arr.append(dict[i]!)
    }
    return arr
  }
}


var nums: [Int?] = [0,3,4,2,1,3,4,2]

var i = 0
var tree = Tree()
var root = tree.setTree(&nums, &i)

var j = 0
var arr: [Int] = tree.setArray(&root, &j)
print(arr)
